gobblin-runtime/src/main/java/gobblin/runtime/Task.java

```diff
@@ -296,9 +359,26 @@ private void processRecord(Object convertedRecord, ForkOperator forkOperator, Ro
       throw new CopyNotSupportedException(convertedRecord + " is not copyable");
     }
 
-    for (int i = 0; i < branches; i++) {
-      if (this.forks.get(i).isPresent() && forkedRecords.get(i)) {
-        this.forks.get(i).get().processRecord(makesCopy ? ((Copyable) convertedRecord).copy() : convertedRecord);
+    // If the record has been successfully put into the queues of every forks
+    boolean allPutsSucceeded = false;
+    // Use an array of primitive boolean type to avoid unnecessary boxing/unboxing
+    boolean[] succeededPuts = new boolean[branches];
+    // Put the record into the record queue of each fork. A put may timeout and return a false, in which
+    // case the put needs to be retried in the next iteration along with other failed puts. This goes on
+    // until all puts succeed, at which point the task moves to the next record.
+    while (!allPutsSucceeded) {
+      allPutsSucceeded = true;
+      for (int i = 0; i < branches && !succeededPuts[i]; i++) {
```

Comment:
It seems this loop ends as soon as succeededPuts[i] = true? Then if there's j > i, succeededPuts[j] = false, j will not be processed?